### PR TITLE
feat(turbo-tasks-fetch): allow specifying a proxy

### PR DESCRIPTION
**Pull request by kwonoj**
### Description

next/font supports to setup proxy.

Closes PACK-2449

### CODE REVIEW

File: crates/turbo-tasks-fetch/src/lib.rs [hunk 1]
```diff
@@ -36,11 +36,32 @@ impl HttpResponseBody {
     }
 }
 
+#[turbo_tasks::value(shared)]
+#[derive(Debug)]
+pub enum ProxyOption {
+    None,
+    Http(String),
+    Https(String),
+}
```
Comment (ForsakenHarmony) [inline crates/turbo-tasks-fetch/src/lib.rs:45]:
```suggestion
#[turbo_tasks::value(shared)]
#[derive(Debug)]
pub enum Proxy {
    Http(String),
    Https(String),
}
```

File: crates/turbo-tasks-fetch/src/lib.rs [hunk 2]
```diff
@@ -36,11 +36,32 @@ impl HttpResponseBody {
...
+
 #[turbo_tasks::function]
-pub async fn fetch(url: Vc<String>, user_agent: Vc<Option<String>>) -> Result<Vc<FetchResult>> {
+pub async fn fetch(
+    url: Vc<String>,
+    user_agent: Vc<Option<String>>,
+    proxy_option: Vc<ProxyOption>,
```
Comment (ForsakenHarmony) [inline crates/turbo-tasks-fetch/src/lib.rs:51]:
```suggestion
    proxy_option: Vc<Option<Vc<ProxyOption>>>,
```

File: crates/turbo-tasks-fetch/src/lib.rs (hunk 1 above)
Comment (ForsakenHarmony) [reply crates/turbo-tasks-fetch/src/lib.rs:45]: Could probably also make this derive `TaskInput` and not have this be a turbo tasks value at all

File: crates/turbo-tasks-fetch/src/lib.rs [hunk 3]
```diff
@@ -36,11 +36,34 @@ impl HttpResponseBody {
...
+
+#[turbo_tasks::value(transparent)]
+pub struct OptionProxyConfig(Option<ProxyConfig>);
+
 #[turbo_tasks::function]
-pub async fn fetch(url: Vc<String>, user_agent: Vc<Option<String>>) -> Result<Vc<FetchResult>> {
+pub async fn fetch(
+    url: Vc<String>,
+    user_agent: Vc<Option<String>>,
+    proxy_option: Vc<OptionProxyConfig>,
```
Comment (ForsakenHarmony) [inline crates/turbo-tasks-fetch/src/lib.rs:53]: does it not work without the newtype?

Comment (kwonoj) [reply crates/turbo-tasks-fetch/src/lib.rs:53]: I got some build error message like cannot be awaited without intofuture trait impl, not sure what's the difference.

Comment (kwonoj) [reply crates/turbo-tasks-fetch/src/lib.rs:53]:
```
error[E0277]: the trait bound `Vc<std::option::Option<ProxyConfig>>: IntoFuture` is not satisfied
  --> crates/turbo-tasks-fetch/src/lib.rs:57:39
   |
57 |     let proxy_option = &*proxy_option.await?;
   |                                      -^^^^^
   |                                      ||
   |                                      |the trait `IntoFuture` is not implemented for `Vc<std::option::Option<ProxyConfig>>`
   |                                      help: remove the `.await`
   |
   = help: the following other types implement trait `IntoFuture`:
             Vc<T>
             &Vc<T>
```

Comment (ForsakenHarmony) [reply crates/turbo-tasks-fetch/src/lib.rs:53]: I think you still have to wrap the inner type in a vc
`Vc<Option<Vc<ProxyConfig>>`

But I guess that makes it a bit more annoying to use

Comment (kwonoj) [reply crates/turbo-tasks-fetch/src/lib.rs:53]: Looks like so. Probably ok to have a new type?

**Review by ForsakenHarmony — APPROVED** [review]

**Pull request merged**